### PR TITLE
[BP-62] Refactor read op, and introduce batchReadOp.

### PR DESCRIPTION
**Pull request by horizonzy**
### Motivation
This is the fourth PR for the batch read(https://github.com/apache/bookkeeper/pull/4051) feature.

Refactor read op, extract ReadOpBase. Introduce batchedReadOp.

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BatchedReadOp.java [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        this.maxCount = maxCount;
+        this.maxSize = maxSize;
+    }
+
+    @Override
+    void initiate() {
+        this.requestTimeNanos = MathUtils.nowInNano();
+        List<BookieId> ensemble = getLedgerMetadata().getEnsembleAt(startEntryId);
+        if (parallelRead) {
+            LOG.info("Batch read unsupported the parallelRead, fail back to sequence read.");
```
Comment (hangc0276) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BatchedReadOp.java:65]: fallback

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BatchedReadOp.java [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+import org.apache.bookkeeper.net.BookieId;
+import org.apache.bookkeeper.proto.BookieProtocol;
+import org.apache.bookkeeper.proto.BookkeeperInternalCallbacks.BatchedReadEntryCallback;
+import org.apache.bookkeeper.proto.checksum.DigestManager;
+import org.apache.bookkeeper.util.ByteBufList;
+import org.apache.bookkeeper.util.MathUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BatchedReadOp extends ReadOpBase implements BatchedReadEntryCallback {
```
Comment (hangc0276) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BatchedReadOp.java:40]: We'd better add a test for BatchedReadOp and PendingReadOp to cover this change.

Comment (horizonzy) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BatchedReadOp.java:40]: The class `TestParallelRead` already covers the PendingReadOp. And I will cover the BatchedReadOp at the next PR. At the next PR, I will introduce the batch read API in LedgerHandle.

**Review by hangc0276 — APPROVED** [review]

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BatchedReadOp.java (hunk 2 above)
Comment (zymap) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BatchedReadOp.java:40]: Do we need a unit test to ensure the bytebuf ref count is correct and the request is handled correctly, like how the ParallelRead test did?

**Comment by hangc0276** [comment]
@merlimat @lhotari @eolivelli Please help take a look, thanks.

**Comment by horizonzy** [comment]
rerun failure checks

**Review by merlimat — APPROVED** [review]

**Pull request merged**